src/importer.cpp

```diff
@@ -0,0 +1,483 @@
+    return resolveComponentImports(model, baseFile, history);
+}
+
+bool Importer::ImporterImpl::resolveImport(const ImportedEntityPtr &importedEntity,
+                                           const std::string &destination,
+                                           const std::string &typeString,
+                                           const std::string &baseFile,
+                                           std::vector<std::tuple<std::string, std::string, std::string>> &history)
+{
+    if (importedEntity->isImport()) {
+        ImportSourcePtr importSource = importedEntity->importSource();
+        auto h = std::make_tuple(destination, importedEntity->importReference(), importSource->url());
+
+        // Check for cyclical dependencies using the entire tuple contents.
+        if (std::find(history.begin(), history.end(), h) != history.end()) {
+            history.emplace_back(h);
+            return false;
+        }
+        history.emplace_back(h);
+
+        // Only parse import if it doesn't already exist in our importer library.
+        if (!importSource->hasModel()) {
+            std::string url = importSource->url();
+            if (mLibrary.count(url) == 0) {
+                url = resolvePath(importSource->url(), baseFile);
+            }
+            if (mLibrary.count(url) == 0) {
+                // If the url has not been resolved into a model in this library, parse it and save.
+                std::ifstream file(url);
+                if (file.good()) {
+                    std::stringstream buffer;
+                    buffer << file.rdbuf();
+                    auto parser = Parser::create();
+                    auto model = parser->parseModel(buffer.str());
+                    importSource->setModel(model);
+
+                    // Save the pair of model and url to the library map.
+                    mLibrary.insert(std::make_pair(url, model));
+
+                    // Save the imported model to a list of external dependencies.
+                    mExternals.emplace_back(std::make_pair(url, importedEntity->importReference()));
+                    return doResolveImports(model, url, history);
+                }
+            } else {
+                // If it has, then pass the previous model instance to the import source.
+                auto model = mLibrary[url];
+                importSource->setModel(model);
+                if (typeString == "component") {
+                    return checkComponentForCycles(model, history);
+                }
+                return checkUnitsForCycles(model, history);
+            }
+        }
+    }
+    return true;
+}
+
+bool Importer::ImporterImpl::resolveComponentImports(const ComponentEntityPtr &parentComponentEntity,
+                                                     const std::string &baseFile,
+                                                     std::vector<std::tuple<std::string, std::string, std::string>> &history)
+{
+    bool noErrors = true;
```

Comment:
I am not a big fan of variables called `noXxx`. Why not simply `bool errors = false;` and then set `errors` to `true`, if needed, and `return !errors;`?

Reply:
I don't like naming variables that shadow internal function names.  The `Importer` class is a `Logger`, so it has `error` (and `warning` etc) functions.

Reply:
I am suggesting `errors`, not `error`. Otherwise, `hasErrors` could be an alternative.